### PR TITLE
Causal effect estimation

### LINKED ISSUE #71: [MB] Week of March 29
- [x] Put together a workflow for estimating causal effects from different causal graphs
   - Deliverable: Opened issue or comment in existing issue with workflow for this effort
   - Estimate: 1 hour
   - Actual: 1 hour

- [x] Run through half of the workflow by midweek
   - Deliverable: New branch with new pushed notebooks with code to be reviewed by Tim and Hassan
   - Estimate: 4 hours
   - Actual: 3 hours

- [x] Finish running through the workflow for estimating causal effects
   - Deliverable: Updated notebook with completed workflow
   - Estimate: 5 hours
   - Actual: 3 hours

- [x] Weekly Meeting:
   -  Estimate: 2 hours
   - Actual: 2 hours

### PR DESCRIPTION
**Pull request by bouzaghrane**
@timothyb0912 @hassanobeid1994 check out a sample implementation of the workflow.

### CODE REVIEW

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2014]: I've only really skipped to the end of your notebook to check out the plots but they're looking right!!

Can you add the following as it may be slightly easier to see than the plot itself?
```suggestion
all_effects =\
    pd.DataFrame({'true_effect': true_effect, 'naive_effect': naive_effect, 'estimated_effect': estimated_effect})
all_effects.describe()
```

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:1184]: This causal graph is very strange. Shouldn't there be edges from
- travel distance to travel time,
- travel distance to travel cost, and
- travel distance to the utility (wtw)?

Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md:1184]: Not really, because X doesn't change here at all. Travel distance for anything involving transit is 0. We can definitely change it to have travel distance be unobserved. Another reason it's not here is because it doesn't enter the utility function in your model.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md]: Here, you're looking at the `naive_effect` as a vector, as in 1 per observation. However, we want `naive_effect` to be once per simulation so `naive_effect = naive_effect.mean)`.

The same goes for your `estimated_effect` and `true_effect`.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2174]: So here you're still storing the model estimation results per simulation and then plotting those later.

We instead want to compute, for each simulation, the `naive_effect`, `estimated_effect`, and `true_effect`.
Again, we want to be sure to compute a scalar value for each of those on each simulation. You should then store these values in an array for later plotting.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md]: We definitely don't want to plot the estimated parameters from each simulation.

We are interested in the causal effects not the estimated parameters per say.

If you create the arrays of naive, estimated, and true causal effects as described in my comment above, those should be plotted here. It should be a single call to seaborns distplot or kdeplot, no need for a loop when plotting.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md:2174]: yea definitely. I still didn't push the notebook containing this and storing the means as well. This loop is incomplete. I will pushing code tomorrow or late this evening.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md]: I agree! The distributions plotted before the loop are the distributions of the causal effects for each observation. Here in the loop we will be taking those and repeating that process N times and then plotting the distribution of the mean. This is just remaining code from notebook 5.0.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md]: Yessir.  The loop isn't finished yet..sorry about that.

**Review by timothyb0912 — CHANGES_REQUESTED** [review]
Hey Amine, I think you're close to being done here.

The major change that's needed is that:
- you're currently thinking of the naive, estimated, and true causal effects as being a vector of values, on each simulation. It's not. They should be scalars per each simulation, and you get the vector by looping over simulations. Again, remember N simulations, each with 1 perturbation, each perturbation giving you a triplet of scalars: the naive, estimated, and true causal effect value for that perturbation.
- At the end, don't store or plot the estimated parameters. Just store and plot the naive, estimated, and true causal effects as you already have. the difference is that the creation of the arrays being plotted happens within the simulation instead before it.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [reply notebooks/working/6.0-mab-estimate-causal-effect.md]: Also, per the last comment in #74 can you use the following causal effect estimate for the naive, estimated, and true estimates?
```suggestion
naive_effect = long_sim_data_naive.loc[long_sim_data_naive['mode_id'].isin([1, 2, 3]), 'naive_probabilities'] - long_sim_data.loc[long_sim_data['mode_id'].isin([1, 2, 3]), 'initial_probabilities']
```

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md:2014]: Done!

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md:2174]: We need to also figure out how to store these models and set up the experiment. I store the necessary causal effect quantities. I'll keep this unresolved until tomorrow and discuss how we want to store all the outputs.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md]: Just pushed a notebook with the graphs we want. @timothyb0912

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md:1184]: @timothyb0912 let me know what you think!

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md]: Why this particular perturbation?

I'm just thinking that the initial interventions that we spoke of (see #5) all relate to reducing travel distance. Either literally (move people closer to their job) or effectively (via electric bikes as they reduce the effective distance for cycling).

Comment (bouzaghrane) [reply notebooks/working/6.0-mab-estimate-causal-effect.md]: @timothyb0912 This was just an example I forgot to clean before I moved on to the loop. Just ignore it. Check out the plots at the bottom. For now the loop only looks at the change in drive alone. I am running it on my laptop right now to plot the estimated effects on all driving modes.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2339]: Hell yes.

These are EXACTLY the results we want.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2008]: Yep yep, these are looking right as well!

**Review by timothyb0912 — APPROVED** [review]
This looks prototype complete to me.

**Comment by bouzaghrane** [comment]
@timothyb0912 sweet! I responded to your comments. Feel free to read my responses and let me know what you think. I'm running the loop right now to estimate the causal effects on all 3 driving modes instead of just drive alone.

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [reply notebooks/working/6.0-mab-estimate-causal-effect.md]: Ah, I see later that you use a perturbation 0.8.

This is cool. You show that both ways (increasing or decreasing travel distance) leads to incorrect causal effect inferences.

👍

**Comment by timothyb0912** [comment]
hmm, for some reason I don't see your responses. Are they really underneath my comments?

**Review by bouzaghrane — COMMENTED** [review]
Adding responses to @timothyb0912 comments.

**Comment by bouzaghrane** [comment]
@timothyb0912 smh I never hit submit review...

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [reply notebooks/working/6.0-mab-estimate-causal-effect.md:1184]: ah yes, good call. That all makes sense!

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [reply notebooks/working/6.0-mab-estimate-causal-effect.md]: no worries!

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2322]: 👍

**Review by timothyb0912 — APPROVED** [review]
It all looks good to me!

**Comment by timothyb0912** [comment]
And yes, we can discuss storage tomorrow and get to work on experiment setup after that!

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2357]: I'm assuming this is a typo that should be removed?

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2289]: @hassanobeid1994 and @bouzaghrane:
I just want to double check that this is indeed the correct computation of the true probabilities, according to the model used to generate the data (as opposed to using the model estimated from the generated data).

It looks like it is, but I remember @hassanobeid1994 that you pointed out in one of our sunday calls that there was an issue where the wrong probabilities were being used. As far as I know, the issue was never recorded in an official comment on github, so we don't have a way of tracking if the issue was correctly and finally resolved.

This is my attempt at creating such a paper trail. @hassanobeid1994, can you:
- review this pull-request one more time,
- respond to this comment to confirm that everything is correct, and
- formally approve this pull-request using the "review changes" button that is available when you click on "commits" or "files changed" from the main pull-request page?

File: notebooks/working/6.0-mab-estimate-causal-effect.md
Comment (timothyb0912) [inline notebooks/working/6.0-mab-estimate-causal-effect.md:2200]: Can you include a comment under each of the `######### Mode Name #######` lines that alerts readers to the effect of using `2 * regression_parameter` in the model for the downstream nodes? In particular, the models for the downstream nodes no longer produce data with the exact means as the real data, but the data is exaggerated to clearly illustrate the causal effect differences that can possibly arise in real life.

**Comment by timothyb0912** [comment]
@bouzaghrane after getting @hassanobeid1994's approval and responding to my other two comments above, I think you will be good to go and should merge this pull request.

**Comment by timothyb0912** [comment]
@bouzaghrane, correct me if I'm wrong, but this PR is no longer relevant right? If so, can you close it?

**Pull request closed without merging**
